### DISCUSSION TITLE
Why does Select and Combobox always display the value instead of the label?

### DISCUSSION THREAD

**Discussion by jrodriguez-sch** [ariakit/ariakit#3944] [Questions]
I'm using the Select and Combobox components to allow users to select options which are based on much more complicated objects. I'm having to take the value and then use `find` to target the specific option chosen, but that doesn't seem to really matter. I'm just pointing out that my use case is more complicated than what I'm asking.

I'm noticing that even though I'm parsing down my objects to only have label and value, whenever an option is selected, the value is the thing that gets displayed when I want to show the label.

For example, I have a list of things like:
```
[
  { id: '1', name: 'Apple', growsOn: 'tree' },
  { id: '2', name: 'Grape', growsOn: 'vine' },
  { id: '3', name: 'Orange', growsOn: 'tree' },
  { id: '4', name: 'Strawberry', growsOn: 'bush' },
  { id: '5', name: 'Watermelon', growsOn: 'vine' },
]
```
and in order to display them in either Select or Combobox, I'm turning them into:
```
[
  { value: '1', label: 'Apple' },
  { value: '2', label: 'Grape' },
  { value: '3', label: 'Orange' },
  { value: '4', label: 'Strawberry' },
  { value: '5', label: 'Watermelon' },
]
```
Users won't care about the ID or know what it means, but we need that on the backend for more accurate processing (names are not always unique).

Whenever one of the options is chosen the ID is shown instead of the label. Shouldn't it show the label since that's what the label is for?

The only workaround seems to just make the label and the value the same thing, but that won't work because there are duplicates (which cannot be fixed). Labels are also not always this contrived, but my use case doesn't matter. IMO, the label is what should be used for display, not the value; especially given that's what the user saw in the list of options and is expecting that to be what displays when the popover closes. What am I missing?

---

Edit: I'm seeing that it's not just about setting selectedValue, but also value as well. If I do `value={selectedValue.label}` then it displays the label I'm after. That wasn't apparent initially.

**Comment by diegohaz** [comment]
You can render anything you want inside `<Select>`:

```jsx
<Select>
  {options.find((option) => option.value === selectedValue)?.label}
</Select>
```

Example: https://ariakit.org/examples/select-item-custom

**Reply by jrodriguez-sch** [reply]
Yea, for select it was relatively easy. It was a bit trickier in Combobox.

Interestingly, when working with Select and using SelectValue, we started getting TS compile errors saying that SelectValue cannot be used as a JSX element. I had to basically render the value manually like you mention vs using that component.

![Screenshot 2024-07-16 at 13 55 18](https://github.com/user-attachments/assets/13788777-8f34-4b7f-9ba3-b6ec0e74809c)

This is how we were using it:
```
<Ariakit.SelectValue>
  {value => {
    const selectedItem = items.find(item => item.value === value)
    return (
      <>
        {/* If the selected item has a left media, use that. 
        If not, use the select's left media. 
        Otherwise, don't show anything. */}
        {!isNil(selectedItem?.leftMedia)
          ? selectedItem.leftMedia
          : !isNil(leftMedia)
          ? leftMedia
          : null}

        {/* Show the placeholder if there is no selected item */}
        <div className={valueStyle(condensed)}>
          {selectedItem ? selectedItem.label : placeholder}
        </div>
      </>
    )
  }}
</Ariakit.SelectValue>
```

**Reply by diegohaz** [reply]
Could you create a bug report issue with a StackBlitz reproduction of that TS error please?
